File: src/common/wasm/zig/argpartition.zig
//! WASM argpartition kernels for all numeric types.
//!
//! Argpartition: fills out[0..N] with indices such that out[kth] is the index
//! of the element that would be at position kth in a sorted array, all indices
//! before kth reference values <= a[out[kth]], and all after reference values
//! >= a[out[kth]].
//!
//! Output is f64 for JS ergonomics. Internally partitions u32 indices in the
//! lower half of the f64 output buffer, then converts in-place via SIMD.
//!
//! Uses quickselect with median-of-three pivot selection.
//! Supported dtypes: f64, f32, i64, u64, i32, u32, i16, u16, i8, u8.
//! Floating-point types treat NaN as greater than all values.

const sc = @import("sorting_common.zig");

/// Helper: partition into u32 indices in the f64 buffer, then convert to f64 in-place.
inline fn argpartitionToF64(comptime T: type, a: [*]const T, out: [*]f64, N: u32, kth: u32) void {
    const out_u32: [*]u32 = @ptrCast(out);
    sc.initIndices(out_u32, N);
    if (N <= 1 or kth >= N) {
        sc.indicesToF64(out_u32, out, N);
        return;
    }
    sc.quickselectIndirect(T, a, out_u32, 0, N - 1, kth);
    sc.indicesToF64(out_u32, out, N);
}

/// Helper: batch slice argpartition into u32, then convert to f64.
inline fn argpartitionSlicesToF64(comptime T: type, a: [*]const T, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    const out_u32: [*]u32 = @ptrCast(out);
    sc.quickselectIndirectSlices(T, a, out_u32, sliceSize, numSlices, kth);
    sc.sliceIndicesToF64(out_u32, out, sliceSize, numSlices);
}

// --- Single-array argpartition (f64 index output) ---

/// Argpartition for f64: indices as f64. NaN sorts to end.
export fn argpartition_f64(a: [*]const f64, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(f64, a, out, N, kth);
}

/// Argpartition for f32: indices as f64. NaN sorts to end.
export fn argpartition_f32(a: [*]const f32, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(f32, a, out, N, kth);
}

/// Argpartition for i64: indices as f64.
export fn argpartition_i64(a: [*]const i64, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(i64, a, out, N, kth);
}

/// Argpartition for u64: indices as f64.
export fn argpartition_u64(a: [*]const u64, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(u64, a, out, N, kth);
}

/// Argpartition for i32: indices as f64.
export fn argpartition_i32(a: [*]const i32, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(i32, a, out, N, kth);
}

/// Argpartition for u32: indices as f64.
export fn argpartition_u32(a: [*]const u32, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(u32, a, out, N, kth);
}

/// Argpartition for i16: indices as f64.
export fn argpartition_i16(a: [*]const i16, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(i16, a, out, N, kth);
}

/// Argpartition for u16: indices as f64.
export fn argpartition_u16(a: [*]const u16, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(u16, a, out, N, kth);
}

/// Argpartition for i8: indices as f64.
export fn argpartition_i8(a: [*]const i8, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(i8, a, out, N, kth);
}

/// Argpartition for u8: indices as f64.
export fn argpartition_u8(a: [*]const u8, out: [*]f64, N: u32, kth: u32) void {
    argpartitionToF64(u8, a, out, N, kth);
}

// --- Batch slice argpartition (f64 index output) ---

/// Argpartition numSlices contiguous f64 slices at kth position. Indices as f64.
export fn argpartition_slices_f64(a: [*]const f64, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(f64, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous f32 slices at kth position. Indices as f64.
export fn argpartition_slices_f32(a: [*]const f32, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(f32, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous i64 slices at kth position. Indices as f64.
export fn argpartition_slices_i64(a: [*]const i64, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(i64, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous u64 slices at kth position. Indices as f64.
export fn argpartition_slices_u64(a: [*]const u64, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(u64, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous i32 slices at kth position. Indices as f64.
export fn argpartition_slices_i32(a: [*]const i32, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(i32, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous u32 slices at kth position. Indices as f64.
export fn argpartition_slices_u32(a: [*]const u32, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(u32, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous i16 slices at kth position. Indices as f64.
export fn argpartition_slices_i16(a: [*]const i16, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(i16, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous u16 slices at kth position. Indices as f64.
export fn argpartition_slices_u16(a: [*]const u16, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(u16, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous i8 slices at kth position. Indices as f64.
export fn argpartition_slices_i8(a: [*]const i8, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(i8, a, out, sliceSize, numSlices, kth);
}

/// Argpartition numSlices contiguous u8 slices at kth position. Indices as f64.
export fn argpartition_slices_u8(a: [*]const u8, out: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    argpartitionSlicesToF64(u8, a, out, sliceSize, numSlices, kth);
}

// --- Tests ---

fn outIdx(out: []const f64, i: usize) usize {
    return @intFromFloat(out[i]);
}

test "argpartition_f64 basic" {
    const testing = @import("std").testing;
    const a = [_]f64{ 3.0, 1.0, 4.0, 1.5, 2.0 };
    var out: [5]f64 = undefined;
    argpartition_f64(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 2.0);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_f64 kth=0" {
    const testing = @import("std").testing;
    const a = [_]f64{ 5.0, 3.0, 1.0, 4.0, 2.0 };
    var out: [5]f64 = undefined;
    argpartition_f64(&a, &out, 5, 0);
    try testing.expectEqual(a[outIdx(&out, 0)], 1.0);
    for (1..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 0)]);
}

test "argpartition_f64 kth=N-1" {
    const testing = @import("std").testing;
    const a = [_]f64{ 5.0, 3.0, 1.0, 4.0, 2.0 };
    var out: [5]f64 = undefined;
    argpartition_f64(&a, &out, 5, 4);
    try testing.expectEqual(a[outIdx(&out, 4)], 5.0);
    for (0..4) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 4)]);
}

test "argpartition_f64 NaN handling" {
    const testing = @import("std").testing;
    const nan = @as(f64, @bitCast(@as(u64, 0x7FF8000000000000)));
    const a = [_]f64{ nan, 2.0, 1.0, nan, 3.0 };
    var out: [5]f64 = undefined;
    argpartition_f64(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 3.0);
    for (0..2) |i| {
        try testing.expect(a[outIdx(&out, i)] <= 3.0);
        try testing.expect(a[outIdx(&out, i)] == a[outIdx(&out, i)]);
    }
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] != a[outIdx(&out, i)]);
}

test "argpartition_i32 basic" {
    const testing = @import("std").testing;
    const a = [_]i32{ 10, -5, 3, 7, -2 };
    var out: [5]f64 = undefined;
    argpartition_i32(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 3);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_u8 basic" {
    const testing = @import("std").testing;
    const a = [_]u8{ 50, 10, 40, 20, 30 };
    var out: [5]f64 = undefined;
    argpartition_u8(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 30);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_i32 already sorted" {
    const testing = @import("std").testing;
    const a = [_]i32{ 1, 2, 3, 4, 5 };
    var out: [5]f64 = undefined;
    argpartition_i32(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 3);
}

test "argpartition_f64 single element" {
    const testing = @import("std").testing;
    const a = [_]f64{42.0};
    var out: [1]f64 = undefined;
    argpartition_f64(&a, &out, 1, 0);
    try testing.expectEqual(out[0], 0.0);
}

test "argpartition_f32 basic" {
    const testing = @import("std").testing;
    const a = [_]f32{ 3.0, 1.0, 4.0, 1.5, 2.0 };
    var out: [5]f64 = undefined;
    argpartition_f32(&a, &out, 5, 2);
    try testing.expectApproxEqAbs(a[outIdx(&out, 2)], 2.0, 1e-5);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_i64 basic" {
    const testing = @import("std").testing;
    const a = [_]i64{ 50, 10, 40, 20, 30 };
    var out: [5]f64 = undefined;
    argpartition_i64(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 30);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_u64 basic" {
    const testing = @import("std").testing;
    const a = [_]u64{ 50, 10, 40, 20, 30 };
    var out: [5]f64 = undefined;
    argpartition_u64(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 30);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_u32 basic" {
    const testing = @import("std").testing;
    const a = [_]u32{ 50, 10, 40, 20, 30 };
    var out: [5]f64 = undefined;
    argpartition_u32(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 30);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_i16 basic" {
    const testing = @import("std").testing;
    const a = [_]i16{ 50, 10, 40, 20, 30 };
    var out: [5]f64 = undefined;
    argpartition_i16(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 30);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_u16 basic" {
    const testing = @import("std").testing;
    const a = [_]u16{ 500, 100, 400, 200, 300 };
    var out: [5]f64 = undefined;
    argpartition_u16(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 300);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_i8 basic" {
    const testing = @import("std").testing;
    const a = [_]i8{ 50, 10, 40, 20, 30 };
    var out: [5]f64 = undefined;
    argpartition_i8(&a, &out, 5, 2);
    try testing.expectEqual(a[outIdx(&out, 2)], 30);
    for (0..2) |i| try testing.expect(a[outIdx(&out, i)] <= a[outIdx(&out, 2)]);
    for (3..5) |i| try testing.expect(a[outIdx(&out, i)] >= a[outIdx(&out, 2)]);
}

test "argpartition_slices_f64 basic" {
    const testing = @import("std").testing;
    const a = [_]f64{ 3.0, 1.0, 2.0, 9.0, 7.0, 8.0 };
    var out: [6]f64 = undefined;
    argpartition_slices_f64(&a, &out, 3, 2, 1);
    // out indices are local (0-based within each slice)
    try testing.expectApproxEqAbs(a[0 + outIdx(&out, 1)], 2.0, 1e-10); // kth=1 of [3,1,2]
    try testing.expectApproxEqAbs(a[3 + outIdx(&out, 4)], 8.0, 1e-10); // kth=1 of [9,7,8]
}

test "argpartition_slices_f32 basic" {
    const testing = @import("std").testing;
    const a = [_]f32{ 3.0, 1.0, 2.0 };
    var out: [3]f64 = undefined;
    argpartition_slices_f32(&a, &out, 3, 1, 1);
    try testing.expectApproxEqAbs(a[outIdx(&out, 1)], 2.0, 1e-5);
}

test "argpartition_slices_i64 basic" {
    const testing = @import("std").testing;
    const a = [_]i64{ 30, 10, 20 };
    var out: [3]f64 = undefined;
    argpartition_slices_i64(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 20);
}

test "argpartition_slices_u64 basic" {
    const testing = @import("std").testing;
    const a = [_]u64{ 30, 10, 20 };
    var out: [3]f64 = undefined;
    argpartition_slices_u64(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 20);
}

test "argpartition_slices_i32 basic" {
    const testing = @import("std").testing;
    const a = [_]i32{ 30, 10, 20 };
    var out: [3]f64 = undefined;
    argpartition_slices_i32(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 20);
}

test "argpartition_slices_u32 basic" {
    const testing = @import("std").testing;
    const a = [_]u32{ 30, 10, 20 };
    var out: [3]f64 = undefined;
    argpartition_slices_u32(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 20);
}

test "argpartition_slices_i16 basic" {
    const testing = @import("std").testing;
    const a = [_]i16{ 30, 10, 20 };
    var out: [3]f64 = undefined;
    argpartition_slices_i16(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 20);
}

test "argpartition_slices_u16 basic" {
    const testing = @import("std").testing;
    const a = [_]u16{ 300, 100, 200 };
    var out: [3]f64 = undefined;
    argpartition_slices_u16(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 200);
}

test "argpartition_slices_i8 basic" {
    const testing = @import("std").testing;
    const a = [_]i8{ 30, 10, 20 };
    var out: [3]f64 = undefined;
    argpartition_slices_i8(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 20);
}

test "argpartition_slices_u8 basic" {
    const testing = @import("std").testing;
    const a = [_]u8{ 200, 50, 100 };
    var out: [3]f64 = undefined;
    argpartition_slices_u8(&a, &out, 3, 1, 1);
    try testing.expectEqual(a[outIdx(&out, 1)], 100);
}

File: src/common/wasm/argpartition.ts
/**
 * WASM-accelerated argpartition (returns indices for quickselect partition).
 *
 * Returns null if WASM can't handle this case.
 */

import {
  argpartition_f64,
  argpartition_f32,
  argpartition_i64,
  argpartition_u64,
  argpartition_i32,
  argpartition_u32,
  argpartition_i16,
  argpartition_u16,
  argpartition_i8,
  argpartition_u8,
  argpartition_slices_f64,
  argpartition_slices_f32,
  argpartition_slices_i64,
  argpartition_slices_u64,
  argpartition_slices_i32,
  argpartition_slices_u32,
  argpartition_slices_i16,
  argpartition_slices_u16,
  argpartition_slices_i8,
  argpartition_slices_u8,
} from './bins/argpartition.wasm';
import {
  wasmMalloc,
  resetScratchAllocator,
  resolveInputPtr,
  scratchCopyIn,
  scratchAlloc,
  getSharedMemory,
  f16InputToScratchF32,
} from './runtime';
import { ArrayStorage } from '../storage';
import type { DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 64;

type ArgpartitionFn = (aPtr: number, outPtr: number, N: number, kth: number) => void;
type SliceArgpartitionFn = (
  aPtr: number,
  outPtr: number,
  sliceSize: number,
  numSlices: number,
  kth: number
) => void;

const kernels: Partial<Record<DType, ArgpartitionFn>> = {
  float64: argpartition_f64,
  float32: argpartition_f32,
  int64: argpartition_i64,
  uint64: argpartition_u64,
  int32: argpartition_i32,
  uint32: argpartition_u32,
  int16: argpartition_i16,
  uint16: argpartition_u16,
  int8: argpartition_i8,
  uint8: argpartition_u8,
  float16: argpartition_f32,
};

const sliceKernels: Partial<Record<DType, SliceArgpartitionFn>> = {
  float64: argpartition_slices_f64,
  float32: argpartition_slices_f32,
  int64: argpartition_slices_i64,
  uint64: argpartition_slices_u64,
  int32: argpartition_slices_i32,
  uint32: argpartition_slices_u32,
  int16: argpartition_slices_i16,
  uint16: argpartition_slices_u16,
  int8: argpartition_slices_i8,
  uint8: argpartition_slices_u8,
  float16: argpartition_slices_f32,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
  int64: BigInt64Array,
  uint64: BigUint64Array,
  int32: Int32Array,
  uint32: Uint32Array,
  int16: Int16Array,
  uint16: Uint16Array,
  int8: Int8Array,
  uint8: Uint8Array,
  float16: Float32Array,
};

/**
 * WASM-accelerated argpartition of contiguous slices.
 * Uses batch kernel when slices are packed contiguously.
 *
 * Note: operates on pre-existing JS buffers, uses scratch.
 */
export function wasmArgpartitionSlices(
  inputData: TypedArray,
  resultData: Int32Array,
  inputSliceOffsets: Int32Array | number[],
  outputSliceOffsets: Int32Array | number[],
  axisSize: number,
  outerSize: number,
  kth: number,
  dtype: DType
): boolean {
  if (axisSize < 2) return false;

  const sliceKernel = sliceKernels[dtype];

  if (
    sliceKernel &&
    inputSliceOffsets[0] === 0 &&
    outerSize > 1 &&
    inputSliceOffsets[1] === axisSize &&
    outputSliceOffsets[0] === 0 &&
    outputSliceOffsets[1] === axisSize
  ) {
    const Ctor = ctorMap[dtype];
    if (!Ctor) return false;
    const outputBytes = resultData.length * 4;

    wasmConfig.wasmCallCount++;
    resetScratchAllocator();

    const inputPtr =
      dtype === 'float16'
        ? f16InputToScratchF32(
            { data: inputData, isWasmBacked: false, wasmPtr: 0, offset: 0 },
            inputData.length
          )
        : scratchCopyIn(inputData as TypedArray);
    const outputPtr = scratchAlloc(outputBytes);

    sliceKernel(inputPtr, outputPtr, axisSize, outerSize, kth);

    const mem = getSharedMemory();
    new Uint8Array(resultData.buffer, resultData.byteOffset, resultData.byteLength).set(
      new Uint8Array(mem.buffer, outputPtr, resultData.byteLength)
    );
    return true;
  }

  // Fallback: per-slice calls
  const kernel = kernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return false;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const outputBytes = resultData.length * 4;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  const inputPtr =
    dtype === 'float16'
      ? f16InputToScratchF32(
          { data: inputData, isWasmBacked: false, wasmPtr: 0, offset: 0 },
          inputData.length
        )
      : scratchCopyIn(inputData as TypedArray);
  const outputPtr = scratchAlloc(outputBytes);

  for (let i = 0; i < outerSize; i++) {
    kernel(
      inputPtr + inputSliceOffsets[i]! * bpe,
      outputPtr + outputSliceOffsets[i]! * 4,
      axisSize,
      kth
    );
  }

  const mem = getSharedMemory();
  new Uint8Array(resultData.buffer, resultData.byteOffset, resultData.byteLength).set(
    new Uint8Array(mem.buffer, outputPtr, resultData.byteLength)
  );

  return true;
}

/**
 * WASM-accelerated argpartition of a contiguous 1D buffer.
 * Returns ArrayStorage of int32 indices or null if WASM can't handle it.
 */
export function wasmArgpartition(a: ArrayStorage, kth: number): ArrayStorage | null {
  if (!a.isCContiguous) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = a.dtype;
  const kernel = kernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  const outBytes = size * 4; // i32 indices

  const outRegion = wasmMalloc(outBytes);
  if (!outRegion) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  let aPtr: number;
  let actualKernel = kernel;

  if (dtype === 'float16') {
    // "Sort floats as integers" trick: transform f16 bit patterns so that
    // signed i16 comparison matches f16 ordering, then use argpartition_i16.
    const f16Bytes = size * 2;
    const mem = getSharedMemory();
    aPtr = scratchAlloc(f16Bytes);
    if (a.isWasmBacked) {
      new Uint8Array(mem.buffer, aPtr, f16Bytes).set(
        new Uint8Array(mem.buffer, a.wasmPtr + a.offset * 2, f16Bytes)
      );
    } else {
      const aData = a.data.subarray(a.offset, a.offset + size) as TypedArray;
      new Uint8Array(mem.buffer, aPtr, f16Bytes).set(
        new Uint8Array(aData.buffer, aData.byteOffset, aData.byteLength)
      );
    }
    const u16View = new Uint16Array(mem.buffer, aPtr, size);
    for (let j = 0; j < size; j++) {
      if (u16View[j]! & 0x8000) u16View[j]! ^= 0x7fff;
    }
    actualKernel = argpartition_i16;
  } else {
    aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, bpe);
  }

  actualKernel(aPtr, outRegion.ptr, size, kth);

  return ArrayStorage.fromWasmRegion(
    Array.from(a.shape),
    'int32',
    outRegion,
    size,
    Int32Array as unknown as new (
      buffer: ArrayBuffer,
      byteOffset: number,
      length: number
    ) => TypedArray
  );
}
